rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -49,28 +71,113 @@
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    public abstract static class CacheExtensionPoint implements ExtensionPoint {
+        public abstract Cache<String,List<FlowNode>> getExecutionCache();
+        public abstract Cache<String, RunExt> getRunCache();
+        public abstract Cache<FlowNode,String> getExecNodeNameCache();
+    }
 
-            if (startTime == 0) {
-                // The node is running and the time has not been marked on it yet.  Return 0 as the duration for now.
-                return 0;
+    // Used in testing where Jenkins is not running yet
+    private static final List<CacheExtension> FALLBACK_CACHES = Arrays.asList(new CacheExtension());
+
+    @Extension
+    public static class CacheExtension extends CacheExtensionPoint {
+
+        protected final Cache<String,List<FlowNode>> executionCache = CacheBuilder.newBuilder().weakValues().maximumSize(100).build();
```

Comment:
@jglick Already addressed two different ways.  First, the whole RunExt gets cached once the run and the run analysis are complete; this execution cache is only used while processing the run. So, it only needs to live for a few seconds in the cache to get the full use (with 10k+ FlowNodes under analysis, about 5-6 seconds on my box).  

Second, in the parent methods that generate the run, we obtain and hold a strong ref to the List while the analysis is progressing - https://github.com/jenkinsci/pipeline-stage-view-plugin/pull/4/commits/50d093a5f7d4aa1591572d2735635ecbb95e8a63#diff-66fea105c6bbbe32760bdb0c2a604c10R336

Worst case, even if it weren't strong-referenced, it would just mean we ran slower because we wouldn't get values from cache.